doc/index.rst

```diff
@@ -8,6 +8,142 @@ lsst.ts.tcpip
 
 Python code to support TCP/IP communication using asyncio.
 
+.. _lsst.ts.tcpip-user_guide:
+
+User Guide
+==========
+
+Here is an overview of how to use this package.
+
+This guide also serves as a very basic tutorial for using `asyncio streams`_,
+and covers a few points I found unclear in that documentation.
+
+In the examples below ``reader`` is an `asyncio.StreamReader` and ``writer`` is an `asyncio.StreamWriter`.
+
+* Use `asyncio.open_connection` to construct a TCP/IP client.
+  For example::
+
+    reader, writer = await asyncio.open_connection(host=..., port=...)
+
+* When you write data, be sure to call drain::
+
+    writer.write(data_bytes)
+    await writer.drain()
+
+* When you read data, check for a closed connection as follows::
+
+    try:
+        data = reader.readuntil(tcpip.TERMINATOR)
+        # or reader.read, tcpip.read_into, etc.
+    catch (asyncio.exceptions.IncompleteReadError, ConnectionResetError):
+        # Connection is closed
+        ...handle the closed connection
+
+  Note that you can only detect a closed connection in the reader;
+  writing to a closed connection does not raise an exception.
+
+* To read and write text data::
+
+    # Reading
+    try:
+        read_bytes = reader.readuntil(tcpip.TERMINATOR)
+        read_text = read_data.decode().strip()
+    catch (asyncio.exceptions.IncompleteReadError, ConnectionResetError):
+        # Connection is closed
+        ...handle the closed connection
+
+    # Writing
+    text_to_write = "some text to write"
+    writer.write(text_to_write.encode() + tcpip.TERMINATOR)
+    await writer.drain()
+
+
+  Note that ``tcpip.TERMINATOR = b"\r\n"``; this is a common way to terminate TCP/IP text.
+
+* To read and write binary data, define a `ctypes.Structure` and use `read_into` and `write_from`.
+  For example::
+
+    class TrivialStruct(ctypes.Structure):
+        _pack_ = 1
+        _fields_ = [
+            ("a_double", ctypes.c_double),
+            ("three_ints", ctypes.c_int64 * 3)
+
+    # Reading
+    data = TrivialStruct()
+    try:
+        await tcpip.read_into(reader, data)
+    catch (asyncio.exceptions.IncompleteReadError, ConnectionResetError):
+        # Connection is closed
+        ...handle the closed connection
+
+    # Writing (write_from does call writer.drain)
+    data = TrivialStruct()
+    data.a_double = 1.2e3
+    data.three_ints[:] = (-1, 2, 3)
+    await write_from(writer, data)
+
+* When you are done with a TCP/IP stream, close the stream writer (you cannot close a stream reader).
+  To close a stream writer, call `close_stream_writer`::
+
+    await close_stream_writer(writer)
```

Comment:
Shouldn't this be `await utils.close_stream_writer(writer)` or at least mention that that function is in the `utils` package?

Reply:
Good point. It should be await tcpip.close_stream_writer. I'll fix it.